### DISCUSSION TITLE
sql functions with named params?

### DISCUSSION THREAD

**Discussion by amirotin** [sqlalchemy/sqlalchemy#8474] [Usage Questions] [answered]
Hi. Is it possible to call sql functions with named params?
For example in postgres there is `make_interval` function with 7 params:
`make_interval(years int DEFAULT 0, months int DEFAULT 0, weeks int DEFAULT 0, days int DEFAULT 0, hours int DEFAULT 0, mins int DEFAULT 0, secs double precision DEFAULT 0.0)`
You can call it as `make_interval(secs => some_column_with_sec)` to make seconds interval based on some column in table, but using `func` i had to write `func.make_interval(0,0,0,0,0,0,table.c.some_column_with_sec)` which is not very convenient.
I didn't find any mention in documentation about using named parameters

**Comment by zzzeek** [comment, marked as answer]
to render that syntax on the backend exactly would be a little more work (still totally doable), however just to get the expediency part of it you can make a [GenericFunction](https://docs.sqlalchemy.org/en/14/core/functions.html?highlight=genericfunction#sqlalchemy.sql.functions.GenericFunction)

```py
from sqlalchemy.sql.functions import GenericFunction
from sqlalchemy import select, func


class make_interval(GenericFunction):
    def __init__(
        self, years=0, months=0, weeks=0, days=0, hours=0, mins=0, secs=0, **kw
    ):
        super().__init__(years, months, weeks, days, hours, mins, secs, **kw)


stmt = select(func.make_interval(weeks=5, hours=0))
print(stmt)
print(
    stmt.compile(
        compile_kwargs={"literal_binds": True}
    )
)

```

which will render as:

```
SELECT make_interval(:make_interval_2, :make_interval_3, :make_interval_4, :make_interval_5, :make_interval_6, :make_interval_7, :make_interval_8) AS make_interval_1
SELECT make_interval(0, 0, 5, 0, 0, 0, 0) AS make_interval_1
```

**Reply by zzzeek** [reply]
to get the full syntax with the `=>` you would build a [@compiles](https://docs.sqlalchemy.org/en/14/core/compiler.html?highlight=compiles#subclassing-guidelines) recipe for it

**Reply by CaselIT** [reply]
Would a feature request issue be appropriate to track this?

I guess the challenge is what to do on backends that don't support it

**Reply by zzzeek** [reply]
right it's not backend agnostic.   i guess if there were something silly like sqlalchemy.dialects.postgresql.kw_func.     the thing is that func() takes keyword arguments `type_`, `packagenames` and I've been keeping that namespace open.

at the same time we could accept keyword args in func, and then if non-PG backend, just render them in order.   kw's after all are ordered in python 3.  it's just what do we do about the existing **kw namespace.    we can of course just "guess".   that always goes well

**Reply by amirotin** [reply]
> right it's not backend agnostic. i guess if there were something silly like sqlalchemy.dialects.postgresql.kw_func. the thing is that func() takes keyword arguments `type_`, `packagenames` and I've been keeping that namespace open.
> 
> at the same time we could accept keyword args in func, and then if non-PG backend, just render them in order. kw's after all are ordered in python 3. it's just what do we do about the existing **kw namespace. we can of course just "guess". that always goes well

Just a small notice - it's not only postgres related. Oracle for example also supports named params with same syntax `=>`

**Reply by CaselIT** [reply]
> at the same time we could accept keyword args in func, and then if non-PG backend, just render them in order. kw's after all are ordered in python 3. it's just what do we do about the existing **kw namespace. we can of course just "guess". that always goes well

Not sure it's possible to do in general without knowing the number of arguments of the function, since given that the kwargs are optional, maybe just the 2nd and 5th arg is passed.

Overall considering the current kw args and the desire of keeping the namespace open, maybe a different call style is required, like a new kwarg that takes a dict, something like `func.make_interval(kws={'secs': 123, 'weeks': 3})`
